### PR TITLE
[improve][pip] PIP-318: Don't retain null-key messages during topic compaction

### PR DESCRIPTION
**Pull request by coderzc**
Motivation & Modifications
Start a PIP: Don't retain null-key messages during topic compaction

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pip/pip-318.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+In order to avoid introducing break changes to release version, we need to add a configuration to control whether to retain null-key messages during topic compaction.
+If the configuration is true, we will retain null-key messages during topic compaction, otherwise, we will not retain null-key messages during topic compaction.
+
+## Public-facing Changes
+
+### Configuration
+
+Add config to broker.conf
+```properties
+compactionRemainNullKey=false
```
Comment (mattisonchao) [inline pip/pip-318.md:28]: `topicCompactionServiceRemainNullKey? `

File: pip/pip-318.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+# Backward & Forward Compatibility
+
+- Make `compactionRemainNullKey=false` default  in the 3.2.0.
+- Cherry-pick it to a branch less than 3.2.0 make `compactionRemainNullKey=true` default.
+- Delete the configuration `compactionRemainNullKey` in 3.3.0 and don't supply an option to retain null-keys.
+
+## Revert
+
+Make `compactionRemainNullKey=true` in broker.conf.
```
Comment (mattisonchao) [inline pip/pip-318.md:39]: and `standalone` also.

File: pip/pip-318.md (hunk 1 above)
Comment (coderzc) [reply pip/pip-318.md:28]: `topicCompactionRemainNullKey` seems to be better, as it requires cherry-picking to the old branch.

File: pip/pip-318.md (hunk 2 above)
Comment (coderzc) [reply pip/pip-318.md:39]: Ok.

File: pip/pip-318.md (hunk 1 above)
Comment (mattisonchao) [reply pip/pip-318.md:28]: Agree.

**Review by mattisonchao — APPROVED** [review]
LGTM.

**Comment by mattisonchao** [comment]
+1 (binding)


Best,
Mattison

**Comment by coderzc** [comment]
Merge this PIP, since the vote has been approved: https://lists.apache.org/thread/36rfmvz5rchgnvqb2wcq4wb64k6st90p

**Pull request merged**